docs/adrs/support-for-public-connections-control-plane-nodes.md

```diff
@@ -0,0 +1,42 @@
+# Support for public connection between servers
+
+Date: 2023-07-18
+
+## Status
+
+Proposed
+
+## Context
+
+A user raised an [issue](https://github.com/k3s-io/k3s/issues/2965) requesting that the embedded etcd of each server should be able to communicate with other servers in clusters outside the LAN connection.
+
+The use case proposed in the issue involves scenarios where servers are not in the same LAN connection but have LAN-like latency between them.
```

Comment:
There is more discussion of this in the thread you linked, but I think we should be VERY cautious around supporting etcd members over non-LAN links. The use case here is for nodes with low-latency non-local connectivity, but if we open the door to running etcd over public networks, people are going to use it and expect it to work even in environments that don't meet that our intended application.

Relevant documentation can be found upstream at:

* https://etcd.io/docs/v3.4/faq/#does-etcd-work-in-cross-region-or-cross-data-center-deployments
  > Since etcd relies on a member quorum for consensus, the latency from crossing data centers will be somewhat pronounced because at least a majority of cluster members must respond to consensus requests. Additionally, cluster data must be replicated across all peers, so there will be bandwidth cost as well.
  > 
  > With longer latencies, the default etcd configuration may cause frequent elections or heartbeat timeouts. See [tuning](https://etcd.io/docs/v3.4/tuning/) for adjusting timeouts for high latency deployments.
* https://etcd.io/docs/v3.4/faq/#what-does-the-etcd-warning-failed-to-send-out-heartbeat-on-time-mean
  > If the etcd cluster is deployed across data centers, long latency between members is expected. For such deployments, tune the heartbeat-interval configuration to roughly match the round trip time between the machines, and the election-timeout configuration to be at least 5 * heartbeat-interval. See [tuning documentation](https://etcd.io/docs/v3.4/tuning/) for detailed information.
  
If we are going to allow this, we'll need to add significant cautions to the documentation, along with some recommendations on how to set the required values appropriately. I would probably want to engage with product management and support to see if this is something we want to actually take a SLA on, or if it will remain experimental and unsupported.

cc @cwayne18 @caroline-suse-rancher

Reply:
Agree with Brad. In both solutions we have for "Distributed hybrid or multicloud cluster" (embedded and tailscale) we left out etcd communication on purpose thus forcing it to be local. Apart from the performance degradation you will see in the cluster, if etcd heartbeats take longer than the election timeout, your cluster will break